### ISSUE TITLE
Some more layout

### ISSUE THREAD

**Issue by circular17** [fredvs/ideU#22]
Configuring fpGUI (overlap)
![ideu_configure_fpgui](https://user-images.githubusercontent.com/9092455/79642021-62757400-819b-11ea-801b-4c0349d70f19.png)

Configuring compiler (window is a bit small)
![ideu_configure_compilers](https://user-images.githubusercontent.com/9092455/79642052-8cc73180-819b-11ea-96ce-7f059fc4bf1f.png)

Configuring shortcuts
![ideu_configureshortcuts](https://user-images.githubusercontent.com/9092455/79642091-d44dbd80-819b-11ea-9197-59ae86d5ef93.png)

Regarding Commandor, icons are small, can we have some setting like 16, 20, 24, 32, 48px? Also comboboxes horizontal width are not scaled to the font.

Otherwise I miss an OK button in Extra settings.

**Comment by fredvs** [comment]
> Configuring fpGUI (overlap)

Ooops, I did forget this one, thanks to note it, I will do it tonight.

> Configuring compiler (window is a bit small)

Could you not resize the window, it is sizable ?
![Capture d’écran_2020-04-18_19-38-25](https://user-images.githubusercontent.com/3421249/79644906-357d8d00-81ac-11ea-82db-a838cbc051e5.png)

> Configuring shortcuts

Well seen, I did forget this one too, I will do it tonight.

> Regarding Commandor, icons are small, can we have some setting like 16, 20, 24, 32, 48px?
 > Also comboboxes horizontal width are not scaled to the font.

This one will need much more work, but, ok, it is in the todo list, for asap.

> Otherwise I miss an OK button in Extra settings.

OK, chef, I will do it tonight.

**Comment by fredvs** [comment]
> Configuring fpGUI (overlap)

> Configuring shortcuts

> Otherwise I miss an OK button in Extra settings.

Done in commit 6cfbbd8

> Regarding Commandor, icons are small, can we have some setting like 16, 20, 24, 32, 48px?

For this I wait for `BRABitmap-MSEgui` because resizing the images of the icons is a good job for him.

;-)



Done in commit 6cfbbd8

**Comment by fredvs** [comment]
Done resizing of buttons in Commandor form (but not yet the icons).
Commit c6226d1

**Comment by circular17** [comment]
Gonna try that

**Comment by circular17** [comment]
Very nice!

> > Configuring compiler (window is a bit small)
> 
> Could you not resize the window, it is sizable ?

Yes but the initial size can be just a bit bigger so that everything fits. Just a few more pixels and no more scrollbars.

About the Commandor, maybe there was a misunderstanding. I was not asking to add spacing between icons (even if that makes it a bit less confusing). I was talking about the comboboxes widths, to select the project, build mode and current file.
![ideU_commandor](https://user-images.githubusercontent.com/9092455/79681812-6944cb00-821d-11ea-8f6f-dab7f2f488cd.png)

Some more things: the tooltip texts and the confirmation dialog text message are small.
![ideU_confirm_dialog_font_size](https://user-images.githubusercontent.com/9092455/79681800-487c7580-821d-11ea-868e-cdbd37a91ebf.png)
![ideU_tooltip](https://user-images.githubusercontent.com/9092455/79681833-a27d3b00-821d-11ea-895f-2cd6bead532c.png)

**Comment by fredvs** [comment]
About the buttons of Commandor, sorry but I dont undestand.
Do you want bigger icons if the size of font is bigger?

If  so, I am obliged to make the icons bigger because with a font size height = 12, it fit perfectly the size of the button.

> I was talking about the comboboxes widths, to select the project, build mode and current file.

Ha ok, I was thinking to keep it little to not make the commandor form too large, but ok, I will do it.
May I ask what font size are you using (so I can see what appends)?

> Some more things: the tooltip texts and the confirmation dialog text message are small.

Ha, ok, so I will create custom tooltip texts and the confirmation dialog because the default one dont allow scaling.
I fear I could do this only asap (because it will take some time).

**Comment by fredvs** [comment]
> to add spacing between icons 

There was not space added between icons!
It seems so because the button have a "flat" look, but the space inside the button is the same.

MSEgui dont do stretching of image (or I did not catch how to do), maybe BGABitmat can do the tick?

**Comment by circular17** [comment]
Don't worry about the look of the buttons they are fine.

Of course one can resize with BGRABitmap. In MSEgui I almost thought there was no resizing but in fact you can do it like that:
```
  stretched := tbitmap.create(bmk_rgb);
  stretched.width := clientwidth; //stretching for the window
  stretched.height := clientheight;
  bmpsource.bitmap.paint(stretched.canvas, makerect(0,0,stretched.width,stretched.height),
                   makerect(0,0,bmpsource.width,bmpsource.height),
                   [al_stretchx,al_stretchy]);
```

**Comment by fredvs** [comment]
Ha, thanks for the tip, always good to have a MSE guru near you!

OK, I did try this but, like I suspected, the result is not good with a bitmap image.
Maybe with a `svg` image but not sure MSEgui can deal with it.

I fear that the best solution is to have 3 different sizes of bitmap for each icon.

**Comment by circular17** [comment]
Ok well there is no hurry for that. It is fine like that. Just the comboboxes are not long enough.

**Comment by fredvs** [comment]
Too late, you put this in my head!
But ok, I will do the comboboxes now and icons asap.

**Comment by circular17** [comment]
I suggest to make a separate issue for the icons, so that we can close this one and have the subject of the icons gently resting:

**Comment by fredvs** [comment]
Ok, in commit 901ff08 the comboboxes are scaled.
But the idea is to have only part of the name of the file, not all 
Did you see that if you click on the combo, all the path is show?

**Comment by circular17** [comment]
Yes, I understand the idea. I will try like that.

**Comment by fredvs** [comment]
Added in last commit: BGRABITMAPDIR global macro.

![Capture d’écran_2020-04-21_01-50-55](https://user-images.githubusercontent.com/3421249/79810410-488e8980-8373-11ea-90dd-aa29492427db.png)

![Capture d’écran_2020-04-21_01-51-15](https://user-images.githubusercontent.com/3421249/79810424-50e6c480-8373-11ea-8b66-975071bc6985.png)

**Comment by circular17** [comment]
Thanks. That's cool we can add * to include all subdirectories I presume?

**Comment by fredvs** [comment]
> Thanks. That's cool we can add * to include all subdirectories I presume?

Of course.

**Comment by circular17** [comment]
Yes I am saying what I am saying because I have seen the pictures.

**Comment by fredvs** [comment]
Ha, ok, so I confirm, you may use things like:

`${BGRABIMAPBIR}*`

`${BGRABIMAPBIR}*/*`

`$../{BGRABIMAPBIR}*`

`$../../{BGRABIMAPBIR}*`

etc...

**Comment by fredvs** [comment]
Note that you may create you custom local macros, in` General Setting | Macros`

**Comment by fredvs** [comment]
Ha, your icon i`deu64.png` gives me a idea.

Maybe stretching a bitmap image from a bigger image to a more little will give better result...
I did try from little bitmap to bigger but the result is pixelated and not nice.

Will see...

**Comment by fredvs** [comment]
@circular17 , about poly-compilation: it is not the same than cross-compilation.
With cross-compilation, you use the same source and same compiler that will produce binary for the OS you want.
With poly-compilation, you use the same source but different compilers, from others OS, 32 or 64 bit.

Of course it is possible also to do cross-compilation in poly-compilation.

**Comment by circular17** [comment]
Yeah maybe

**Comment by fredvs** [comment]
> Yeah maybe

Huh, maybe what, stretching or poly-compilation?

**Comment by fredvs** [comment]
About bitmap things, I am not pro at all, so your advice is highly appreciated.

If you see a better way to render stretched icons, or if you think it is better to create retouched images in some size, etc,... I dont know what is better,

**Comment by circular17** [comment]
Well the best is to use SVG or LZP and stretch them

**Comment by circular17** [comment]
Yep

**Comment by fredvs** [comment]
Yep and aaargh (because now lot of work to convert and tune the png to svg).

What do you use for conversion, Inkscape?

**Comment by fredvs** [comment]
OK, I will begin with the debugger icon, convert in svg, refactor all and try to load it with TBRABitmap.

Huh, by the way, do you have a MSEgui demo to show how to load image using BGRABitmap and MSEgui?

**Comment by circular17** [comment]
> Yep and aaargh (because now lot of work to convert and tune the png to svg).

I know. You have my sympathy.

> What do you use for conversion, Inkscape?

LazPaint of course. LZP can be vectorial now.

> Huh, by the way, do you have a MSEgui demo to show how to load image using BGRABitmap and MSEgui?

That's how it is done in LazPaint (for now applies only to the fill interface in the toolbar for pen, back and outline):
```
procedure LoadToolbarImage(AImages: TImageList; AIndex: integer; AFilename: string);
var
  iconImg: TBGRALazPaintImage;
  iconFlat: TBGRABitmap;
begin
  iconImg := TBGRALazPaintImage.Create;
  iconImg.LoadFromResource(AFilename);
  iconImg.Resample(AImages.Width,AImages.Height,rmFineResample,rfBestQuality);
  iconFlat := TBGRABitmap.Create(iconImg.Width,iconImg.Height);
  iconImg.Draw(iconFlat,0,0);
  if AImages.Count < AIndex then
    AImages.Replace(AIndex, iconFlat.Bitmap,nil)
  else
    AImages.Add(iconFlat.Bitmap,nil);
  iconFlat.Free;
  iconImg.Free;
end;  
```

**Comment by fredvs** [comment]
Ha, ok, I will study this deep.

But now it is the occasion to jump into icon layout.

For example, I would like to use the svg that is in attachment.
Would it work for a canvas of a button face?

[bugnew.svg.zip](https://github.com/fredvs/ideU/files/4511496/bugnew.svg.zip)

**Comment by fredvs** [comment]
> LazPaint of course. LZP can be vectorial now.

Ha, ok, perfect!

**Comment by fredvs** [comment]
>  LazPaint of course. LZP can be vectorial now.

Just downloaded last version of LaZPaint.
WOW, OOTB (I like version without install), very nice.
But I did try to load the svg of previous attachment without luck.
Inskspace do it ok.

**Comment by circular17** [comment]
> > LazPaint of course. LZP can be vectorial now.
> 
> Just downloaded last version of LaZPaint.
> WOW, OOTB (I like version without install), very nice.

Yeah I like to see the binary files and all of the program.

> But I did try to load the svg of previous attachment without luck.
> Inkspace does it ok.

In fact, it is not a real SVG, it is an SVG with an image in it. I suppose raster image is not supported yet in current SVG implementation BGRABitmap.

Anyway when I am talking about using LZP format, I am not talking about importing SVG in it but drawing using LazPaint vector shapes.

If you want to load an SVG file, you can instead use TBGRASVG to load it and do a stretch draw on TBGRABitmap.Canvas2d.

**Comment by circular17** [comment]
Indeed raster image is not supported yet. I've added an issue:
https://github.com/bgrabitmap/bgrabitmap/issues/99

**Comment by fredvs** [comment]
OK.

I did test with image 100x100 png pixels and stretch if until 16x16.
Not bad at all.

I will do it for the 3 most important icons : compile, debug on/off and play.
For the others, first I will wait for your comments.

Many thanks for your tips.

**Comment by fredvs** [comment]
@circular17 : If you dont see any problems, maybe you can close the issue.
When all the issues are fixed, I will release ideU 2.1.0.

**Comment by circular17** [comment]
There are some more layout issues.

When I compile, the commandor is shrunk vertically. Maybe related to big font size?

![ideu_commandor_compile](https://user-images.githubusercontent.com/9092455/80112308-40a43480-8581-11ea-936f-0c19a088bf0f.png)

The procedure list has a small font:
![image](https://user-images.githubusercontent.com/9092455/80112547-83660c80-8581-11ea-89d0-eff7d7b212a0.png)

Same for code beautifier:
![image](https://user-images.githubusercontent.com/9092455/80112629-9bd62700-8581-11ea-977b-499f6e462eab.png)

By the way the filename is centered but does not fit. Maybe align right?

Message dialog still have small font:
![image](https://user-images.githubusercontent.com/9092455/80112836-d8098780-8581-11ea-8f0f-1c7026d7ee2e.png)

**Comment by fredvs** [comment]
> When I compile, the commandor is shrunk vertically. Maybe related to big font size?

Hum, that was not compiled with last commit of ideU (icons are changed).
Anyway, the Commandor form is sizable and people may customize different layout.

![80112308-40a43480-8581-11ea-936f-0c19a088bf0f](https://user-images.githubusercontent.com/3421249/80115173-b4940c00-8584-11ea-9bc3-4ea12a78fe7f.png)

Also when you choose a layout, the default layouts in` /layout` are done for font size 12.
So if you use a different size, you should adapt your new layout.

Once it is done, just save the project (menu `Project-Save as`) with the new layout in the directory `/layout` and choose a name for it.
It does not matter what was the original project, `ideU` will use only the layout data of the project.

Maybe in a future version we could imagine to adapt the layout-list with the font used.

> The procedure list has a small font:
> Same for code beautifier:

Oops, yes ok, I will do it.

> Message dialog still have small font:

Yes, I know but this is lot of work because it is part of main `MSEgui` and needs (imho) to be more customizable.

What can be done now is a custom dialog message for "`Confirmation Delete row?`".
But for all the other MSEgui system dialog messages, sorry but only asap and for the next version of ideU.

**Comment by fredvs** [comment]
Did you see the docking feature of MSEgui?

In the right top corner there is a little circle, it enable-disable docking child. 

![dock1](https://user-images.githubusercontent.com/3421249/80119808-77cb1380-858a-11ea-8ec6-006edb0467b0.png)

Click on it, new icons of docking will appear:

![dock2](https://user-images.githubusercontent.com/3421249/80119857-8adde380-858a-11ea-8289-cb05de931b82.png)

Now you may use the grip right pink panel to dock-undock the window.

![dock3](https://user-images.githubusercontent.com/3421249/80120052-c1b3f980-858a-11ea-864b-ebe00cd0436a.png)

**Comment by circular17** [comment]
> Hum, that was not compiled with last commit of ideU (icons are changed).

Ah oui j'avais remplacé l'exécutable.

> Anyway, the Commandor form is sizable and people may customize different layout.
> See the thin grip panel hatched, click on it and maintain the button while resizing.

Yeah I know that. Problem is that if I adjut it when it is not compiling, when I compile it shrinks and the message is not visible anymore.

> > Message dialog still have small font:
> 
> Yes, I know but this is lot of work because it is part of main `MSEgui` and needs (imho) to be more customizable.

Ok well that can wait.

**Comment by circular17** [comment]
> Did you see the docking feature of MSEgui?
> 
> In the right top corner there is a little circle, it enable-disable docking child.

Ok. Those icons are way too small.

**Comment by fredvs** [comment]
> Ok. Those icons are way too small.

OK, I will see if it is possible to do it now.

**Comment by fredvs** [comment]
> Yeah I know that. Problem is that if I adjut it when it is not compiling, when I compile it shrinks and > the message is not visible anymore.

Sorry but this one I cannot reproduce.
Once you have resized the window, it should remain like it is even when compiling...
Or I miss something.

**Comment by fredvs** [comment]
> Ok. Those icons are way too small.

OK, one in commit 77c7f18

**Comment by fredvs** [comment]
> The procedure list has a small font:
> Same for code beautifier:

OK done in commit: a967380

**Comment by fredvs** [comment]
> Message dialog still have small font:

Fixed in commit 2f21a9c for ALL message dialog forms, including Delete row.

![Capture d’écran_2020-04-24_02-35-28](https://user-images.githubusercontent.com/3421249/80162730-627ad700-85d4-11ea-9a97-3599fd49b141.png)

![Capture d’écran_2020-04-24_02-35-54](https://user-images.githubusercontent.com/3421249/80162741-6870b800-85d4-11ea-8419-dc14d309efe6.png)

**Comment by circular17** [comment]
> > Yeah I know that. Problem is that if I adjut it when it is not compiling, when I compile it shrinks and > the message is not visible anymore.
> 
> Sorry but this one I cannot reproduce.
> Once you have resized the window, it should remain like it is even when compiling...
> Or I miss something.

Ok I think I understand. When the compiler messages are shown at the bottom, it uses some of the vertical space. And as the commandor had not a fixed height, it was reduced proportionally during the compilation.

If I go into the layout icons and select "Fix size" for the commandor then this doesn't happen anymore.

**Comment by circular17** [comment]
It is all fine now. Congrats!

Just one thing left about the filename alignment in code beautifier:
![image](https://user-images.githubusercontent.com/9092455/80187787-8eb44900-8610-11ea-8216-4d5d87c900d5.png)

The pink area at the top.

**Comment by fredvs** [comment]
> The pink area at the top.

OK, it will show only the file name, not the complete path.
OK for you?

**Comment by fredvs** [comment]
> Just one thing left about the filename alignment in code beautifier:

Done with commit 5b6fd5f

**Comment by fredvs** [comment]
> Some more things: the tooltip texts ...

OK, fixed for hints too in commit a17f569.

![hintscaled](https://user-images.githubusercontent.com/3421249/80262212-0f188f80-868d-11ea-93cb-13ff0ab87062.png)
